### ISSUE TITLE
reinterpret crashes between float and pointer on 0.62

### ISSUE THREAD

**Issue by chethega** [JuliaLang/julia#26053]
```
versioninfo()
#Julia Version 0.6.2
#Commit d386e40 (2017-12-13 18:08 UTC)

reinterpret(Float64, C_NULL)

LLVM ERROR: Cannot select: 0x564c43af2820: f64 = truncate 0x564c43af03a0
  0x564c43af03a0: i64,ch = CopyFromReg 0x564c41d2de70, Register:i64 %vreg1
    0x564c438c4440: i64 = Register %vreg1
In function: julia_reinterpret_62749

reinterpret(Ptr{UInt8}, 0.3)

LLVM ERROR: Cannot select: 0x55f14d77bd50: i64 = truncate 0x55f14d5dbe70
  0x55f14d5dbe70: f64,ch = CopyFromReg 0x55f14ba3e040, Register:f64 %vreg1
    0x55f14d5db7e0: f64 = Register %vreg1
In function: julia_reinterpret_62710
```
This is fixed on 0.7. One can work-around by reinterpreting via `UInt64` or similar.

This appeared in context of reinterpreting structs that correspond to C unions, see [https://discourse.julialang.org/t/dealing-with-complex-c-structure/9045](https://discourse.julialang.org/t/dealing-with-complex-c-structure/9045); so it is not entirely silly to reinterpret between Float64 and pointers.

**Comment by chethega** [comment]
Some more details:

```
@code_llvm reinterpret(Ptr{UInt8}, 0.3)

define i8* @julia_reinterpret_62740(i8**, double) #0 !dbg !5 {
top:
  %2 = inttoptr double %1 to i8*
  ret i8* %2
}
@code_native reinterpret(Ptr{UInt8}, 0.3)
#crash!
```
If I understand [https://llvm.org/docs/LangRef.html#inttoptr-to-instruction](https://llvm.org/docs/LangRef.html#inttoptr-to-instruction) correctly, then the llvm code is invalid; hence no wonder that it crashes on compilation to native code. Rather surprising that no optimization pass crashes and burns on it.

I think the fix in 0.7 was accidential:
```
@code_llvm reinterpret(Ptr{Int}, 0.3)

; Function reinterpret
; Location: essentials.jl:332
define i64 @julia_reinterpret_66001(%jl_value_t addrspace(10)*, double) {
top:
  %2 = bitcast double %1 to i64
  ret i64 %2
}
```
So I would expect that the correct variant on 0.6 should first `%2 = bitcast double %1 to i64` and then `%3 = inttoptr i64 %2 to i8*` and  `ret i8* %3` (I don't understand the apparent change in the calling convention here; that is, why we return an `i64` on 0.7).

**Comment by Keno** [comment]
> I don't understand the apparent change in the calling convention here; that is, why we return an i64 on 0.

We've switched to codegening all non-float bitstypes as integers, essentially for reasons like the one in this issue.

**Comment by chethega** [comment]
I take the closure of this issue as "won't fix" for julia 0.6?

In this case I'll just leave the work-around/fix from discourse here, for posterity if someone else encounters the the problem:
```
if sizeof(Ptr{Int})==8
    Base.reinterpret(::Type{Ptr{T}}, x::Float64) where T = reinterpret(Ptr{T}, reinterpret(UInt64, x))
    Base.reinterpret(::Type{Float64}, x::Ptr) = reinterpret(Float64, reinterpret(UInt64, x))
elseif sizeof(Ptr{Int})==4
    #untested on 32 bit
    Base.reinterpret(::Type{Ptr{T}}, x::Float32) where T = reinterpret(Ptr{T}, reinterpret(UInt32, x))
    Base.reinterpret(::Type{Float32}, x::Ptr) = reinterpret(Float32, reinterpret(UInt32, x))
end
```
